sympy/combinatorics/fp_groups.py

```diff
@@ -1003,4 +1018,423 @@ def first_in_class(C, Y=[]):
+                    C.P[mu][A_dict[x]] = v
+                    C.P[nu][A_dict_inv[x]] = v**-1
+
+
+# Pg 350, section 2.5.1 from [2]
+def elimination_technique_1(C):
+    rels = list(C.reidemeister_relators)
+    # the shorter relators are examined first so that generators selected for
+    # elimination will have shorter strings as equivalent
+    rels.sort(reverse=True)
+    gens = list(C.schreier_generators)
+    redundant_gens = {}
+    contained_gens = []
+    # examine each relator in relator list for any generator occuring exactly
+    # once
+    next_i = False
+    for i in range(len(rels) -1, -1, -1):
+        rel = rels[i]
+        # don't look for a new generator occuring once in relator which
+        # has already found to posses a
+        for gen in redundant_gens:
+            gen_sym = gen.array_form[0][0]
+            if any([gen_sym == r[0] for r in rel.array_form]):
+                next_i = True
+                break
+        if next_i:
+            next_i = False
+            continue
+        for j in range(len(gens) - 1, -1, -1):
+            gen = gens[j]
+            if rel.generator_count(gen) == 1 and gen not in contained_gens:
+                k = rel.exponent_sum(gen)
+                gen_index = rel.index(gen**k)
+                bk = rel.subword(gen_index + 1, len(rel))
+                fw = rel.subword(0, gen_index)
+                redundant_gens[gen] = (bk*fw)**(-1*k)
+                contained_gens.extend((bk*fw).contains_generators())
+                del rels[i]; del gens[j]
+                break
+    # eliminate the redundant generator from remaing relators
+    for i, gen in product(range(len(rels)), redundant_gens):
+        rels[i] = rels[i].eliminate_word(gen, redundant_gens[gen])
+    rels.sort()
+    try:
+        rels.remove(C.schreier_free_group.identity)
+    except ValueError:
+        pass
+    C.reidemeister_relators = set(rels)
+    C.schreier_generators = set(gens)
+
+# Pg 350, section 2.5.2 from [2]
+def elimination_technique_2(C):
+    """
+    This technique eliminates one generator at a time. Heuristically this
+    seems superior in that we may select for elimination the generator with
+    shortest equivalent string at each stage.
+
+    >>> from sympy.combinatorics.free_group import free_group
+    >>> from sympy.combinatorics.fp_groups import FpGroup, coset_enumeration_r, reidemeister_relators, define_schreier_generators, elimination_technique_2
+    >>> F, x, y = free_group("x, y")
+    >>> f = FpGroup(F, [x**3, y**5, (x*y)**2]); H = [x*y, x**-1*y**-1*x*y*x]
+    >>> C = coset_enumeration_r(f, H)
+    >>> C.compress(); C.standardize()
+    >>> define_schreier_generators(C)
+    >>> reidemeister_relators(C)
+    >>> elimination_technique_2(C)
+    ([y_1, y_2], [y_2**-3, y_2*y_1*y_2*y_1*y_2*y_1, y_1**2])
+
+    """
+    rels = list(C.reidemeister_relators)
+    rels.sort(reverse=True)
+    gens = list(C.schreier_generators)
+    for i in range(len(gens) - 1, -1, -1):
```

Comment:
Yes.

On 07/04/2016 10:27 AM, Kalevi Suominen wrote:
>
> In sympy/combinatorics/fp_groups.py 
> <https://github.com/sympy/sympy/pull/11295#discussion_r69408413>:
>
> > +    >>> from sympy.combinatorics.free_group import free_group
> > +    >>> from sympy.combinatorics.fp_groups import FpGroup, coset_enumeration_r, reidemeister_relators, define_schreier_generators, elimination_technique_2
> > +    >>> F, x, y = free_group("x, y")
> > +    >>> f = FpGroup(F, [x**3, y**5, (x*y)**2]); H = [x*y, x**-1*y**-1*x*y*x]
> > +    >>> C = coset_enumeration_r(f, H)
> > +    >>> C.compress(); C.standardize()
> > +    >>> define_schreier_generators(C)
> > +    >>> reidemeister_relators(C)
> > +    >>> elimination_technique_2(C)
> > +    ([y_1, y_2], [y_2**-3, y_2*y_1*y_2*y_1*y_2*y_1, y_1**2])
> > +
> > +    """
> > +    rels = list(C.reidemeister_relators)
> > +    rels.sort(reverse=True)
> > +    gens = list(C.schreier_generators)
> > +    for i in range(len(gens) - 1, -1, -1):
>
> Maybe |len(rels) - 1|?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11295/files/1327a1b039a327c531e305bc3663f54362dd9c90#r69408413>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8bHZu1Ch3b70raGsuvmq62bJAqt9ks5qSJKmgaJpZM4I-QZs>.
>